### PR TITLE
Always discover blacklisted folders to avoid data loss when modifying selectivesync list.

### PR DESCRIPTION
**Pull request by allexzander**

### CODE REVIEW

**Review by claucambra — APPROVED** [review]

**Comment by claucambra** [comment]
@allexzander would this be a fix for https://github.com/nextcloud/desktop/issues/1238 ?

**Comment by allexzander** [comment]
> @allexzander would this be a fix for #1238 ?

hmm, it might, but not 100% sure, were you able to reproduce it?

File: test/testlocaldiscovery.cpp [hunk 1]
```diff
@@ -17,6 +17,87 @@ class TestLocalDiscovery : public QObject
...
+        fakeFolder.localModifier().insert("more-big-files-wont-fit/bigfile_A.data", 800);
+        fakeFolder.localModifier().insert("more-big-files-wont-fit/bigfile_B.data", 800);
+
+        const auto remoteQuota = 600;
+        QObject parent;
+        fakeFolder.setServerOverride([&](QNetworkAccessManager::Operation op, const QNetworkRequest &request, QIODevice *outgoingData) -> QNetworkReply * {
+            Q_UNUSED(outgoingData)
+            if (op == QNetworkAccessManager::PutOperation) {
+                if (request.rawHeader("OC-Total-Length").toInt() > remoteQuota) {
+                    return new FakeErrorReply(op, request, &parent, 507);
```
Comment (mgallien) [inline test/testlocaldiscovery.cpp:50]: you can directly use `fakeFolder`, no need to create an extra parent object

File: test/testlocaldiscovery.cpp [hunk 2]
```diff
@@ -17,6 +17,87 @@ class TestLocalDiscovery : public QObject
...
+                }
+            }
+            return nullptr;
+        });
+
+        QVERIFY(!fakeFolder.syncOnce());
+
+        fakeFolder.syncEngine().journal()->setSelectiveSyncList(SyncJournalDb::SelectiveSyncBlackList, {"big-files-wont-fit/", "more-big-files-wont-fit/"});
+        fakeFolder.syncEngine().setLocalDiscoveryOptions(LocalDiscoveryStyle::DatabaseAndFilesystem, {"big-files-wont-fit/", "more-big-files-wont-fit/"});
```
Comment (mgallien) [inline test/testlocaldiscovery.cpp:59]: this is somehow wrong or do I misunderstand the need for those lines ?
it looks like you force the sync engine to have some specific behavior that would not be teh case of a running client

Comment (mgallien) [reply test/testlocaldiscovery.cpp:59]: this is the behavior of the selective sync dialog
so this is expected when changing the list of selective sync folders

File: test/testlocaldiscovery.cpp (hunk 1 above)
Comment (allexzander) [reply test/testlocaldiscovery.cpp:50]: @mgallien no, `fakeFolder `doesn't work, it is not `QObject`

**Review by mgallien — APPROVED** [review]

**Pull request merged**